### PR TITLE
nipype_cmd - run any Interface from a command line

### PR DESCRIPTION
**Pull request by chrisfilo**
nipype_cmd --help

    usage: nipype_cmd [-h] module interface

    Nipype interface runner

    positional arguments:
      module      Module name
      interface   Interface name

    optional arguments:
      -h, --help  show this help message and exit

nipype_cmd nipype.interfaces.nipy

    Available Interfaces:
	    SpaceTimeRealigner
	    Similarity
	    ComputeMask
	    FitGLM
	    EstimateContrast
	    FmriRealign4d

nipype_cmd nipype.interfaces.nipy ComputeMask --help

    usage: nipype_cmd nipype.interfaces.nipy ComputeMask [-h] [--M M] [--cc CC]
                                                         [--ignore_exception IGNORE_EXCEPTION]
                                                         [--m M]
                                                         [--reference_volume REFERENCE_VOLUME]
                                                         mean_volume

    Run ComputeMask

    positional arguments:
      mean_volume           mean EPI image, used to compute the threshold for the
                            mask

    optional arguments:
      -h, --help            show this help message and exit
      --M M                 upper fraction of the histogram to be discarded
      --cc CC               Keep only the largest connected component
      --ignore_exception IGNORE_EXCEPTION
                            Print an error message instead of throwing an
                            exception in case the interface fails to run
      --m M                 lower fraction of the histogram to be discarded
      --reference_volume REFERENCE_VOLUME
                            reference volume used to compute the mask. If none is
                            give, the mean volume is used.

nipype_cmd nipype.interfaces.nipy ComputeMask /tmp/mean_volume.nii

    brain_mask = /tmp/mean_mask.nii.gz

### CODE REVIEW

**Comment by chrisfilo** [comment]
@Garyfallidis I'm gonna clean this up today. Any comments/requests?

**Comment by chrisfilo** [comment]
@Garyfallidis here's a dipy example:

	$ bin/nipype_cmd nipype.interfaces.dipy TensorMode --help
	usage: nipype_cmd nipype.interfaces.dipy TensorMode [-h]
	                                                    [--out_filename OUT_FILENAME]
	                                                    bvals bvecs in_file
	
	Run TensorMode
	
	positional arguments:
	  bvals                 The input b-value text file
	  bvecs                 The input b-vector text file
	  in_file               The input 4D diffusion-weighted image file
	
	optional arguments:
	  -h, --help            show this help message and exit
	  --out_filename OUT_FILENAME
	                        The output filename for the Tensor mode image

**Comment by Garyfallidis** [comment]
Hi @chrisfilo. How can I help? If I remember correctly the idea was to have a common interface in Dipy and Nipype so we can write scripts in Dipy and have them available in Nipype without having the dependency of Nipype in Dipy. Am I correct?

**Comment by chrisfilo** [comment]
nipype_cmd is written in a very generic way right now - it can turn any Nipype Interface into a command line script. Interfaces it operates on do not have to eve nbe part of nipype package. You could do:

	nipype_cmd nipype.interfaces.dipy TensorMode
	
but also:

	nipype_cmd dipy.interfaces TensorMode
	
if dipy was installed and provided Nipype Interfaces. Does this functionality help you in your use cases?

**Comment by chrisfilo** [comment]
Ready for review. Tests fail only on python 2.6. They will start passing after #1069 will be merged ;)

File: doc/users/index.rst [hunk 1]
```diff
@@ -37,6 +37,7 @@
    saving_workflows
    spmmcr
    mipav
+   nipypecmd
```
Comment (satra) [inline doc/users/index.rst]: this file is missing.

Comment (chrisfilo) [inline doc/users/index.rst]: fixed

File: nipype/utils/nipype_cmd.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    interface = None
+    if parser and module and function:
+        __import__(module)
+        interface = getattr(sys.modules[module],function)()
+
+        for k,v in interface.inputs.items():
+            if hasattr(v, "mandatory") and v.mandatory:
+                parser.add_argument(k, help=v.desc)
+            else:
+                parser.add_argument("--%s"%k, dest=k,
```
Comment (satra) [inline nipype/utils/nipype_cmd.py]: do you want to add type information here from traits? then that will help reduce the code in run_instance

**Comment by satra** [comment]
also see here:

https://github.com/nipy/nipype/blob/master/examples/fmri_ants_openfmri.py#L1042

how to add default values and types

**Comment by satra** [comment]
it would also be sweet to add nipype memcache options to this! perhaps that's a sprint item.

**Pull request merged**